File: binary_search/BinarySearch.go
package main

import (
  "fmt"
)

func main(){
  haystack := []int{2, 4, 5, 7, 8, 9, 10}
  needle := 8
  binary_search(haystack, needle)
}

func binary_search(haystack []int, needle int) bool{
  lo := 0
  hi := len(haystack)

  for lo < hi {
    m := lo + (hi - lo)/2
    v := haystack[m]
    if v == needle {
      fmt.Println("guilty as charged")
      fmt.Printf("needle is %d\n", v)
      return true
    } else if v > needle {
      hi = m
    } else {
      lo = m + 1
    }
  }
  fmt.Println("false verdict")
  return false
}

File: c-kata/binary_search.c
#include "main.h"

/**
 * binary_search - implementation of binary search algorithm
 * @haystack - array of ints
 * @size- size of array
 * @needle - number to find
 *
 * Return: bool
 */


void main()
{
  int haystack[] = {3, 5, 7, 12, 17};
  int needle = 12;
  binary_search(haystack, 5, needle);
}

bool binary_search(int *haystack, int size, int needle)
{
  int lo = 0;
  int hi = size;

  do
  {
    int m = lo + (hi - lo)/2;
    int v = haystack[m];

    if (v == needle)
    {
      printf("verdict is true, needle is %d", v);
      return true;
    } else if (v > needle) {
      hi = m;
    } else {
      lo = m + 1;
    }
  } while (lo < hi);

  printf("verdict is false");
  return false;
}
